File: TDDMicroExercises.ProposedSolution/Javascript/telemetry-system/telemetry-diagnostic-controls.js

TelemetryDiagnosticControls = function (telemetryClientConnection, telemetryDataChannel) {

	this._diagnosticChannelConnectionString = function () { return '*111#'; };
	this._diagnosticInfo = '';
	this._telemetryClientConnection = telemetryClientConnection;
	this._telemetryDataChannel = telemetryDataChannel;
};

TelemetryDiagnosticControls.Create = function () {
	return new TelemetryDiagnosticControls(new TelemetryClientConnection(), new TelemetryDataChannel());
};

TelemetryDiagnosticControls.prototype = {

	readDiagnosticInfo: function() {
		return this._diagnosticInfo;
	},

	writeDiagnosticInfo: function(newValue) {
		this._diagnosticInfo = newValue;
	},

	checkTransmission: function() {

		this._diagnosticInfo = '';

		this._telemetryClientConnection.disconnect();

		var retryLeft = 3;
		while (this._telemetryClientConnection.onlineStatus() === false && retryLeft > 0) {
			this._telemetryClientConnection.connect(this._diagnosticChannelConnectionString);
			retryLeft -= 1;
		}

		if (this._telemetryClientConnection.onlineStatus() === false) {
			throw 'Unable to connect';
		}

		this._telemetryDataChannel.send(TelemetryDataChannel.diagnosticMessage());
		this._diagnosticInfo = this._telemetryDataChannel.receive();
	}
};

File: Java/TelemetrySystem/src/main/java/tddmicroexercises/telemetrysystem/TelemetryDiagnostics/TelemetryDiagnosticControls.java
package tddmicroexercises.telemetrysystem.TelemetryDiagnostics;

import tddmicroexercises.telemetrysystem.TelemetryClient.ITelemetryClient;
import tddmicroexercises.telemetrysystem.TelemetryClient.TelemetryClient;
import tddmicroexercises.telemetrysystem.TelemetryConnection.Connection;
import tddmicroexercises.telemetrysystem.TelemetryConnection.TelemetryConnection;

public class TelemetryDiagnosticControls {
    private final String DiagnosticChannelConnectionString = "*111#";

    private final ITelemetryClient telemetryClient;
    private final Connection telemetryConnection;
    private String diagnosticInfo = "";

    public TelemetryDiagnosticControls() {
        telemetryClient = new TelemetryClient();
        telemetryConnection = new TelemetryConnection();
    }

    public String getDiagnosticInfo() {
        return diagnosticInfo;
    }

    public void setDiagnosticInfo(String diagnosticInfo) {
        this.diagnosticInfo = diagnosticInfo;
    }

    public void checkTransmission() throws Exception {
        diagnosticInfo = "";

        telemetryConnection.disconnect();

        int retryLeft = 3;
        while (telemetryConnection.getConnectionStatus() == false && retryLeft > 0) {
            telemetryConnection.connect(DiagnosticChannelConnectionString);
            retryLeft -= 1;
        }

        if (telemetryConnection.getConnectionStatus() == false) {
            throw new Exception("Unable to connect.");
        }

        telemetryClient.send(TelemetryClient.DIAGNOSTIC_MESSAGE);
        diagnosticInfo = telemetryClient.receive();
    }
}